src/manifold/src/manifold.cpp

```diff
@@ -767,4 +830,104 @@ Manifold Manifold::TrimByPlane(glm::vec3 normal, float originOffset) const {
+                        const double precision) {
+  const int len = vertProps.size() / numProp;
+  // FIXME: Should there be a hull specific error?
+  if (len < 4) return Invalid();
+  std::vector<glm::vec3> pts(len);
+  for (int i = 0; i < len; i++) {
+    const int j = i * numProp;
+    pts[i] = {vertProps[j], vertProps[j + 1], vertProps[j + 2]};
+  }
+  const auto trip = NonCollinearTriple(pts, precision);
+  const auto plane = PlaneOfPts(pts[trip.x], pts[trip.y], pts[trip.z]);
+  const auto [d, dIsAboveTrip] = NonCoplanar(pts, plane, precision);
+  if (d < 0) return Invalid();  // All points coplanar.
+  const int a = trip.x;
+  const int b = dIsAboveTrip ? trip.z : trip.y;
+  const int c = dIsAboveTrip ? trip.y : trip.z;
+  std::vector<glm::ivec3> triangles;
+  std::vector<glm::vec4> planes;
+  std::unordered_set<glm::ivec2, IVec2Hash> halfEdges{};
+  std::vector<int> dropped;
+  std::vector<bool> kept;
+  const auto AddTri = [&pts, &triangles, &planes, &dropped, &kept](int a, int b,
+                                                                   int c) {
+    auto tri = glm::ivec3(c, b, a);
+    auto plane = PlaneOfPts(pts[a], pts[b], pts[c]);
+    if (dropped.size() > 0) {
+      const int idx = dropped[dropped.size() - 1];
+      triangles[idx] = tri;
+      planes[idx] = plane;
+      kept[idx] = true;
+      dropped.pop_back();
+    } else {
+      triangles.push_back(tri);
+      planes.push_back(plane);
+      kept.push_back(true);
+    }
+  };
+  AddTri(a, b, c);
+  AddTri(d, b, a);
+  AddTri(c, d, a);
+  AddTri(b, d, c);
+  for (int i = 0; i < len; i++) {
+    if (i == a || i == b || i == c || i == d) continue;  // skip starting points
+    // collect half edges of triangles that are in conflict with the points at
+    // idx, pruning the conflicting triangles and their planes in the process
+    for (int j = 0; j < triangles.size(); j++) {
+      if (kept[j] && DistanceToPlane(pts[i], planes[j]) > precision) {
+        halfEdges.insert({triangles[j].x, triangles[j].z});
+        halfEdges.insert({triangles[j].z, triangles[j].y});
+        halfEdges.insert({triangles[j].y, triangles[j].x});
+        // mark triangle/plane as dropped (slot to be reused)
+        dropped.push_back(j);
+        kept[j] = false;
+      }
+    }
+    // form new triangles with the outer perimeter (horizon) of the set of
+    // conflicting triangles and the point at idx
+    for (auto e : halfEdges) {
+      if (halfEdges.erase({e.y, e.x}) == 0) AddTri(e.x, e.y, i);
+    }
+    halfEdges.clear();
+  }
+  MeshGL mesh;
+  mesh.numProp = numProp;
+  mesh.vertProperties = vertProps;
+  mesh.triVerts.reserve((triangles.size() - dropped.size()) * 3);
+  for (int i = 0; i < triangles.size(); i++) {
+    if (!kept[i]) continue;
+    mesh.triVerts.push_back(triangles[i].x);
+    mesh.triVerts.push_back(triangles[i].y);
+    mesh.triVerts.push_back(triangles[i].z);
+  }
+  return Manifold(mesh);
```

Comment:
I think we should probably not use MeshGL in the internal, as conversion between Manifold and MeshGL is pretty costly.